pkg/codegen/hcl2/model/type.go

```diff
@@ -80,27 +80,54 @@ func assignableFrom(dest, src Type, assignableFromImpl func() bool) bool {
 	return assignableFromImpl()
 }
 
+type cacheEntry struct {
+	kind  ConversionKind
+	diags lazyDiagnostics
+}
+
+type conversionCache map[Type]cacheEntry
+
 func conversionFrom(dest, src Type, unifying bool, seen map[Type]struct{},
+	cache conversionCache,
 	conversionFromImpl func() (ConversionKind, lazyDiagnostics),
 ) (ConversionKind, lazyDiagnostics) {
 	if dest.Equals(src) || dest == DynamicType {
 		return SafeConversion, nil
 	}
 
+	if c, ok := cache[src]; ok {
```

Comment:
Looks like we have a data race flagged in CI https://github.com/pulumi/pulumi/actions/runs/14854224463/job/41703729364?pr=19422#step:41:499

Is this just an issue in the tests since we run the cases in parallel?

Reply:
Yeah, I thought there wasn't any parallelism here, but making it safe seems like a good idea either way.  I changed this to use a `gsync.Map` instead, which should work around this.